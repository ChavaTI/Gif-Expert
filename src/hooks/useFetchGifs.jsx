import { useState, useEffect } from 'react';

export const useFetchGifs = (category) => {
  const [state, setState] = useState({
    images: [],
    loading: true
  });

  useEffect(() => {
    const getGifs = () => {
      const url = `https://api.giphy.com/v1/gifs/search?q=${encodeURI(
        category
      )}&limit=10&api_key=JPU3y1rQZlxJnnGfKadNjHIRkWjX14XG`;
      fetch(url)
        .then((req) => req.json())
        .then((res) => res.data)
        .then((gifs) =>
          gifs.map((gif) => ({
            id: gif.id,
            title: gif.title,
            url: gif.images?.downsized_medium.url
          }))
        )
        .then((imgObject) => {
          setState({
            images: [...imgObject],
            loading: false
          });
        })
        .catch((error) => error);
    };
    getGifs();
  }, [category]);
  return state;
};
