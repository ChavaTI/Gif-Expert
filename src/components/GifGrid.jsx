import React from 'react';
import GifGridItem from './GifGridItem';
import useFetchGifs from '../hooks/useFetchGifs';
import PropTypes from 'prop-types';

const GifGrid = ({ category }) => {
  const { images, loading } = useFetchGifs(category);
  return (
    <>
      <h4>{category}</h4>
      {loading ? (
        'Loading...'
      ) : (
        <div className="card-grid">
          {images.map((image) => (
            <GifGridItem key={image.id} {...image} />
          ))}
        </div>
      )}
    </>
  );
};

GifGrid.propTypes = {
  category: PropTypes.string.isRequired
};

export default GifGrid;
