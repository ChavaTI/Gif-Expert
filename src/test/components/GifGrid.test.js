import '@testing-library/jest-dom';
import React from 'react';
import { shallow } from 'enzyme';
import GifGrid from '../../components/GifGrid';
import useFetchGifs from '../../hooks/useFetchGifs';
jest.mock('../../hooks/useFetchGifs');

describe('GifGrid tests', () => {
  it('should render correctly', () => {
    useFetchGifs.mockReturnValue({
      images: [],
      loading: true
    });
    let category = 'test';
    let wrapper = shallow(<GifGrid category={category} />);
    expect(wrapper).toMatchSnapshot();
  });
  it('should show items when useFetchGifs hook load the data', () => {
    useFetchGifs.mockReturnValue({
      images: [
        {
          id: 'vPLWd7f9rDc9a',
          title: 'prince zuko avatar GIF',
          url:
            'https://media4.giphy.com/media/vPLWd7f9rDc9a/giphy.gif?cid=e78c1e51fw1h92p9pqq0zoxh5hafdb0aefh2i1gdskjvfda1&rid=giphy.gif&ct=g'
        }
      ],
      loading: false
    });
    let category = 'test';
    let wrapper = shallow(<GifGrid category={category} />);
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('p').exists()).toBe(false);
    expect(wrapper.find('GifGridItem').exists()).toBe(true);
  });
  it('should have the same number of GifGridItem and images downloaded', () => {
    const images = [
      {
        id: 'vPLWd7f9rDc9a',
        title: 'prince zuko avatar GIF',
        url:
          'https://media4.giphy.com/media/vPLWd7f9rDc9a/giphy.gif?cid=e78c1e51fw1h92p9pqq0zoxh5hafdb0aefh2i1gdskjvfda1&rid=giphy.gif&ct=g'
      },
      {
        id: 'vPLWd7f9rDc9a',
        title: 'prince zuko avatar GIF',
        url:
          'https://media4.giphy.com/media/vPLWd7f9rDc9a/giphy.gif?cid=e78c1e51fw1h92p9pqq0zoxh5hafdb0aefh2i1gdskjvfda1&rid=giphy.gif&ct=g'
      }
    ];
    useFetchGifs.mockReturnValue({
      images,
      loading: false
    });
    let category = 'test';
    let wrapper = shallow(<GifGrid category={category} />);
    expect(wrapper.find('GifGridItem').lenght).toBe(images.lenght);
  });
});
