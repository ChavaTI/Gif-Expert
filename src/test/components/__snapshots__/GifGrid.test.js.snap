// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GifGrid tests should render correctly 1`] = `
<Fragment>
  <h4>
    test
  </h4>
  Loading...
</Fragment>
`;

exports[`GifGrid tests should show items when useFetchGifs hook load the data 1`] = `
<Fragment>
  <h4>
    test
  </h4>
  <div
    className="card-grid"
  >
    <GifGridItem
      id="vPLWd7f9rDc9a"
      key="vPLWd7f9rDc9a"
      title="prince zuko avatar GIF"
      url="https://media4.giphy.com/media/vPLWd7f9rDc9a/giphy.gif?cid=e78c1e51fw1h92p9pqq0zoxh5hafdb0aefh2i1gdskjvfda1&rid=giphy.gif&ct=g"
    />
  </div>
</Fragment>
`;
